pypicloud/storage/gcs.py

```diff
@@ -0,0 +1,154 @@
+""" Store packages in GCS """
+import calendar
+import logging
+import posixpath
+import time
+from contextlib import contextmanager
+from hashlib import md5
+from urllib import urlopen, quote
+
+from boto.gs.key import Key
+import gcs_oauth2_boto_plugin
+import boto.gs
+from pyramid.httpexceptions import HTTPFound
+from pyramid.settings import asbool
+
+from .base import IStorage
+from pypicloud.models import Package
+from pypicloud.util import parse_filename, getdefaults
+
+
+LOG = logging.getLogger(__name__)
+
+SUPPORTED_CALLING_FORMATS = {
+    'SubdomainCallingFormat': boto.s3.connection.SubdomainCallingFormat,
+    'VHostCallingFormat': boto.s3.connection.VHostCallingFormat,
+    'OrdinaryCallingFormat': boto.s3.connection.OrdinaryCallingFormat,
+    'ProtocolIndependentOrdinaryCallingFormat':
+        boto.s3.connection.ProtocolIndependentOrdinaryCallingFormat
+}
+
+
+class GCSStorage(IStorage):
```

Comment:
There's a lot of code duplication from the `S3Storage` class. If it's possible, it may be better to subclass `S3Storage` directly and only overwrite the behavior that you need to change. Again, the [CloudFront PR](https://github.com/mathcamp/pypicloud/pull/71/files) is a good template to go from.